rest-api/dao/participant_counts_over_time_service.py

```diff
@@ -176,70 +192,163 @@ def get_enrollment_status_sql(self, filters_sql_ps, filters_sql_p):
-                           CASE WHEN
-                                LEAST(
-                                    COALESCE(sample_order_status_1ed04_time, '3000-01-01'),
-                                    COALESCE(sample_order_status_2ed10_time, '3000-01-01'),
-                                    COALESCE(sample_order_status_1ed10_time, '3000-01-01'),
-                                    COALESCE(sample_order_status_1sal_time, '3000-01-01'),
-                                    COALESCE(sample_order_status_1sal2_time, '3000-01-01')
-                                    ) = '3001-01-01' THEN NULL
-                                ELSE LEAST(
-                                    COALESCE(sample_order_status_1ed04_time, '3000-01-01'),
-                                    COALESCE(sample_order_status_2ed10_time, '3000-01-01'),
-                                    COALESCE(sample_order_status_1ed10_time, '3000-01-01'),
-                                    COALESCE(sample_order_status_1sal_time, '3000-01-01'),
-                                    COALESCE(sample_order_status_1sal2_time, '3000-01-01')
-                                    )
-                           END)
-                END) day
-               FROM participant_summary ps
-              %(filters_ps)s
-           GROUP BY day) full
-             ON c2.day = full.day) day_sums, calendar
-          WHERE calendar.day >= :start_date
-            AND calendar.day <= :end_date
-          GROUP BY calendar.day
-          ORDER BY calendar.day;
-      """ % {'filters_ps': filters_sql_ps, 'filters_p': filters_sql_p}
+                       LEFT OUTER JOIN participant_summary ps ON p.participant_id = ps.participant_id
+                %(filters_p)s) a left join
+               (select participant_id,
+                      (
+                          CASE
+                            WHEN enrollment_status>=2 and ps.consent_for_electronic_health_records = 1 THEN DATE(COALESCE(ps.consent_for_electronic_health_records_time, '3000-01-01'))
+                            ELSE Date('3000-01-01') END
+                          ) member_day,
+                       case when enrollment_status=3 then
+                       DATE(
+                         GREATEST(
+                           GREATEST(
+                             COALESCE(sign_up_time, '1000-01-01'),
+                             COALESCE(consent_for_electronic_health_records_time, '1000-01-01'),
+                             COALESCE(questionnaire_on_the_basics_time, '1000-01-01'),
+                             COALESCE(questionnaire_on_lifestyle_time, '1000-01-01'),
+                             COALESCE(questionnaire_on_overall_health_time, '1000-01-01'),
+                             COALESCE(physical_measurements_finalized_time, '1000-01-01')
+                               ),
+                           LEAST(
+                             COALESCE(sample_status_1ed04_time, '3000-01-01'),
```

Comment:
If the only difference between this SQL and the SQL below is sample_status_* instead of sample_order_status_*, we should have a shared SQL template and just substitute in one or the other, rather than having everything else in two places.

Reply:
Actually, I try to do this. There is another different place that make me give up. For the one use sample status time, when I try to get the  core_day, I use enrollment_status=3; For the other one, I use enrollment_status>=2, because for some records with order time, the enrollment status is still 2, but we want to count it as core, so that's why I need it as enrollment_status>=2.